drivers/clock_control/clock_control_nrf2_hsfll.c

```diff
@@ -0,0 +1,211 @@
+/*
+ * Copyright (c) 2024 Nordic Semiconductor ASA
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT nordic_nrf_hsfll
+
+#include "clock_control_nrf2_common.h"
+LOG_MODULE_DECLARE(clock_control_nrf2, CONFIG_CLOCK_CONTROL_LOG_LEVEL);
+
+/* TODO: add support for other HSFLLs */
+BUILD_ASSERT(DT_NUM_INST_STATUS_OKAY(DT_DRV_COMPAT) == 1,
+	     "multiple instances not supported");
+
+#ifdef CONFIG_NRFS_HAS_DVFS_SERVICE
+#include <ld_dvfs_handler.h>
+
+#define UPDATE_TIMEOUT_US K_USEC(1000)
+
+/* TODO: move to devicetree */
+static const uint32_t oppoint_freq[] = { MHZ(64), MHZ(128), MHZ(320) };
+static const uint32_t max_frequency = MHZ(320);
+
+struct hsfll_dev_data {
+	STRUCT_CLOCK_CONFIG(hsfll, ARRAY_SIZE(oppoint_freq)) clk_cfg;
+	struct k_timer timer;
+};
+
+static void freq_setting_applied_cb(enum dvfs_frequency_setting new_setting)
+{
+	ARG_UNUSED(new_setting);
```

Comment:
This part looks invalid. `freq_setting_applied_cb()` can be called unsolicited, when other CPUs are scaling their frequencies and voltages. This function in this CPU should configure HSFLL to apply `new_setting`, even if it is higher than the last requested setting.

It can be fixed in another PR. But at least a comment like `TODO` would be good.

Reply:
And I'm missing the actual code modifying the frequency. Is it somewhere within this file?

Reply:
> And I'm missing the actual code modifying the frequency. Is it somewhere within this file?

It's handled in [ld_dvfs](https://github.com/zephyrproject-rtos/zephyr/blob/5d87cd654fbe2d430bcdaf39c666683abe65f290/modules/hal_nordic/nrfs/dvfs/ld_dvfs.c#L256).

Reply:
> This part looks invalid. `freq_setting_applied_cb()` can be called unsolicited, when other CPUs are scaling their frequencies and voltages.

Yep. I'm reworking this.